src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp

```diff
@@ -0,0 +1,198 @@
+        phi_boundary_correction,
+
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_spacetime_metric_int,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& v_zero_int,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_plus_int,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_minus_int,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& normal_times_v_plus_int,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& normal_times_v_minus_int,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>&
+        constraint_gamma2_v_spacetime_metric_int,
+    const tnsr::a<DataVector, 3, Frame::Inertial>& char_speeds_int,
+
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_spacetime_metric_ext,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>& v_zero_ext,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_plus_ext,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>& v_minus_ext,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>&
+        minus_normal_times_v_plus_ext,
+    const tnsr::iaa<DataVector, Dim, Frame::Inertial>&
+        minus_normal_times_v_minus_ext,
+    const tnsr::aa<DataVector, Dim, Frame::Inertial>&
+        constraint_gamma2_v_spacetime_metric_ext,
+    const tnsr::a<DataVector, 3, Frame::Inertial>& char_speeds_ext) const
+    noexcept {
+  const size_t num_pts = char_speeds_int[0].size();
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>,
+                       ::Tags::TempScalar<2>, ::Tags::TempScalar<3>,
+                       ::Tags::TempScalar<4>, ::Tags::TempScalar<5>,
+                       ::Tags::TempScalar<6>, ::Tags::TempScalar<7>>>
+      buffer(num_pts);
+  DataVector& weighted_lambda_spacetime_metric_int =
+      get(get<::Tags::TempScalar<0>>(buffer));
+  weighted_lambda_spacetime_metric_int = step_function(-char_speeds_int[0]);
+  DataVector& weighted_lambda_spacetime_metric_ext =
+      get(get<::Tags::TempScalar<1>>(buffer));
+  weighted_lambda_spacetime_metric_ext = -step_function(char_speeds_ext[0]);
+
+  DataVector& weighted_lambda_zero_int =
+      get(get<::Tags::TempScalar<2>>(buffer));
+  weighted_lambda_zero_int = step_function(-char_speeds_int[1]);
+  DataVector& weighted_lambda_zero_ext =
+      get(get<::Tags::TempScalar<3>>(buffer));
+  weighted_lambda_zero_ext = -step_function(char_speeds_ext[1]);
+
+  DataVector& weighted_lambda_plus_int =
+      get(get<::Tags::TempScalar<4>>(buffer));
+  weighted_lambda_plus_int = step_function(-char_speeds_int[2]);
+  DataVector& weighted_lambda_plus_ext =
+      get(get<::Tags::TempScalar<5>>(buffer));
+  weighted_lambda_plus_ext = -step_function(char_speeds_ext[2]);
+
+  DataVector& weighted_lambda_minus_int =
+      get(get<::Tags::TempScalar<6>>(buffer));
+  weighted_lambda_minus_int = step_function(-char_speeds_int[3]);
+  DataVector& weighted_lambda_minus_ext =
+      get(get<::Tags::TempScalar<7>>(buffer));
+  weighted_lambda_minus_ext = -step_function(char_speeds_ext[3]);
+
```

Comment:
A comment clarifying how things change vs the equations in the help text because the normals point in opposite directions would be helpful, as would some comments regarding the signs. Since this empirically is stable, I presume it's all correct here, but I find it nontrivial to follow the signs here, to make sure they all are correct.

Reply:
I'm adding the SpECTRE form into the dox instead